File: ruby/lib/runapi/topaz/resources/video_upscales.rb
# frozen_string_literal: true

module RunApi
  module Topaz
    module Resources
      class VideoUpscales
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/topaz/video_upscales"

        RESPONSE_CLASS = Types::VideoUpscaleResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedVideoUpscaleResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "model is required" unless param(params, :model) == Types::VIDEO_UPSCALE_MODEL
          raise Core::ValidationError, "video_url is required" unless param(params, :video_url)

          factor = param(params, :upscale_factor)
          return unless factor && !Types::VIDEO_UPSCALE_FACTORS.include?(factor.to_s)

          raise Core::ValidationError, "upscale_factor must be one of: #{Types::VIDEO_UPSCALE_FACTORS.join(", ")}"
        end
      end
    end
  end
end

File: js/src/resources/video-upscales.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { TaskCreateResponse, VideoUpscaleParams, VideoUpscaleResponse } from '../types';

const ENDPOINT = '/api/v1/topaz/video_upscales';

export class VideoUpscales {
  constructor(private readonly http: HttpClient) {}

  async run(params: VideoUpscaleParams, options?: RequestOptions & PollingOptions): Promise<VideoUpscaleResponse> {
    const { id } = await this.create(params, options);
    return pollUntilComplete<VideoUpscaleResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
  }

  async create(params: VideoUpscaleParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body: compactParams(params),
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<VideoUpscaleResponse> {
    return this.http.request<VideoUpscaleResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}
